nav2_experimental_rl/turtlebot3_env.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (c) 2019 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from time import sleep
+from threading import Thread
+
+import rclpy
+from rclpy.node import Node
+from rclpy.duration import Duration
+
+import numpy as np
+import math
+import random
+import pandas
+import parameters
+
+from geometry_msgs.msg import Twist
+from sensor_msgs.msg import LaserScan
+from std_srvs.srv import Empty
+from std_msgs.msg import String
+from gazebo_msgs.srv import GetEntityState, SetEntityState
+
+class TurtlebotEnv():
+    def __init__(self):
+        node_ = rclpy.create_node('turtlebot3_env')
+        self.act = 0
+        self.done = False
+        self.actions = [[parameters.LINEAR_FWD_VELOCITY, parameters.ANGULAR_FWD_VELOCITY],
+                        [parameters.LINEAR_STR_VELOCITY, parameters.ANGULAR_VELOCITY],
+                        [parameters.LINEAR_STR_VELOCITY, -parameters.ANGULAR_VELOCITY]]
+        self.bonous_reward = 0
+
+        self.collision = False
+        self.collision_tol = 0.125
+        self.laser_scan_range = [0] * 360
+        self.states_input = [3.5] * 8
+        self.zero_div_tol = 0.01
+        self.range_min = 0.0
+
+        self.pub_cmd_vel = node_.create_publisher(Twist, 'cmd_vel')
+        self.sub_scan = node_.create_subscription(LaserScan, 'scan', self.scan_callback)
+
+        self.reset_simulation = node_.create_client(Empty, 'reset_simulation')
+        self.reset_world = node_.create_client(Empty, 'reset_world')
+        self.unpause_proxy = node_.create_client(Empty, 'unpause_physics')
```

Comment:
These proxys arent being used. This is probably important so that the simulation pauses between updates to simulate actual N hz control, the differential drive plugin in gazebo has no timeout entry so the last command will make the robot go in that cmd_vel forever or until a new command is processed. For a mobile robot its less noticable than a manipulator that would drop when it didn't receive commands fast enough but it should be done so that the training looks closer to reality in running on hardware.

I.e. if you want control efforts at 20hz, every time you send a command, let it forward simulate 1/20 of a second, the pause, run through control efforts, unpause, send command, let it forward simulate for 1/20 of a second, and repeat. This will look in simulation time closer to reality even if unintuitive.

Reply:
Example snippet from my code:
```
        self.unpause_proxy()
        tStart = rospy.Time.now() #ros time so if simulation running at not 1x speed it will be simulated correctly
        self.pub_cmd_vel.publish(vel_cmd)
        while not rospy.is_shutdown():
          should_exit = rospy.Time.now() - tStart > self.control_duration
          if should_exit:
            break
        self.pause_proxy()
```

Reply:
@SteveMacenski Actually I put pause/unpause proxies there for this intention and also being able to run at x times real-time. But at the time that I implemented this code, I couldn't get current time from rclpy.  Maybe now time.now() is available in rclpy.  Let me check and if it is I will update the code.  Thanks.

Reply:
OK. Another piece of advise I can give you from having done this is that you need to make sure you unpause it on exit conditions so that new sensor readings are able to get received on the first iteration of the new session. Since we send a stop command on collision/goal we're not at risk of getting erroneous simulation artifacts.

Reply:
Will do this in another PR once the rclpy have the missing functionality.